lib/puppet/configurer/plugin_handler.rb

```diff
@@ -41,8 +41,7 @@ def download_plugins(environment)
       result += locales_downloader.evaluate
     end
 
-
-    Puppet::Util::Autoload.reload_changed
+    Puppet::Util::Autoload.reload_changed(Puppet.lookup(:current_environment))
```

Comment:
Seems like `reload_changed` should be called after we evaluate facts that we just downloaded, since facts can determine which environment we switch to.